lib/semantics/scope.h

```diff
@@ -89,6 +89,8 @@ class Scope {
   const Symbol *GetSymbol() const;
   const Scope *GetDerivedTypeParent() const;
 
+  // It is only safe to call name() for kind of scopes for which GetSymbol
```

Comment:
See above.  The real problem is that `name()` looks like an accessor and clients shouldn't have to worry about crashing.

`std::optional<SourceName> GetName() const { ... }` would be better.

Reply:
I agree, though I think the return type should be `const SourceName *` because other member functions of Scope with optional return values return pointers (e.g. `symbol()`, `GetDerivedTypeParent()`).

Clients can get the old behavior with `DEREF(scope.GetName())` when they know the scope has a name.

Reply:
In this particular case, a `SourceName` is small, and only value semantics are needed, I think.

Reply:
OK, I think the most important here is to be consistent throughout  f18 regarding how "maybe" `SourceName` are returned. Here is a quick survey of the existing code base:

- 4 functions are returning `const SourceName *` (8 bytes, but it is a pointer)

- 1 is returning `std::optional<SourceName>` (24 bytes)

Should I transform everything to `std::optional<SourceName>` or go with the existing majority and use `const SourceName *` everywhere ?


Detailed Survey:


`const SourceName *GetSubmoduleParent(const parser::Program &);`
https://github.com/flang-compiler/f18/blob/a1e26df148c4d480178aa8b765aad4ceaaa5eae2/lib/semantics/mod-file.cc#L39
`const SourceName *MessageHandler::currStmtSource();`
https://github.com/flang-compiler/f18/blob/a1e26df148c4d480178aa8b765aad4ceaaa5eae2/lib/semantics/resolve-names.cc#L100
`const SourceName *BaseVisitor::currStmtSource();`
https://github.com/flang-compiler/f18/blob/a1e26df148c4d480178aa8b765aad4ceaaa5eae2/lib/semantics/resolve-names.cc#L140
`const SourceName *WithPassArg::passName();`
https://github.com/flang-compiler/f18/blob/009db27bfaca2a84b0795c0d977c68cd8b57ce88/lib/semantics/symbol.h#L186


`std::optional<SourceName> DerivedTypeDetails::GetParentComponentName();`
https://github.com/flang-compiler/f18/blob/009db27bfaca2a84b0795c0d977c68cd8b57ce88/lib/semantics/symbol.h#L233

Reply:
Are you sure that all of those APIs have value semantics?  Some might be returning optional references encoded as pointers.  It might be hard to tell.

It's okay if you don't change all of these APIs, just this one.  We can change other APIs to value semantics later when visiting that code.

Reply:
Any of these could return `std::optional<SourceName>` or `const SourceName *`. I think it is good for them to be consistent, and also consistent with `symbol()`, which returns `const Symbol *` because it can't return by value.

I don't see any benefit to using `std::optional<SourceName>`.

Reply:
`symbol()` is returning an optional reference to an object, not a value.  `GetName()` is returning an optional value, not a reference.  The benefit of making a distinction between references and values is that references to objects raise questions about ownership and expiration; values results don't.

Reply:
Comment treated in latest changes:
I have changed all `SourceName` to use value semantics by replacing all `SourceName *` by `std::optional<SourceName>`. This includes both some function return types and some class member types. I also changed argument types to be consistent.

`SourceName Scope::name()` is changed to `std::optional<SourceName> Scope::GetName()` and related call sites modified.